### ISSUE TITLE
Implement more RTCP stuff

### ISSUE THREAD

**Issue by ibc** [versatica/mediasoup#29]
**mediasoup** will not relay RTCP packets but consume and generate them. This is, **mediasoup** will act as a media endpoint by negotiating and managing RTCP with each client.

* Docs: https://github.com/ibc/mediasoup/wiki/RTCP

**Comment by ibc** [comment]
Some interesting issues/posts here:

* [Chrome/Firefox/Licode REMB issue](https://groups.google.com/forum/#!topic/discuss-webrtc/pXETZPysNZs)

**Comment by ibc** [comment]
Lof of RTCP stuff done by @jmillan in #41

**Comment by Globik** [comment]
Rtsp is an analog of rtcp?

**Comment by ibc** [comment]
NACK processed! Habemus RTP retransmission.

**Comment by ibc** [comment]
Among with other tasks, we must also define which exact RTCP feedback parameters mediasoup supports. For example, current Chrome supports the following values:

```
a=rtcp-fb:100 ccm fir
a=rtcp-fb:100 nack
a=rtcp-fb:100 nack pli
a=rtcp-fb:100 goog-remb
a=rtcp-fb:100 transport-cc
```

**Comment by jmillan** [comment]
Below are the RTCP types we are supporting.

For Feedback type "ack" or possitive acknowledgment:
- ack rpsi (bypassed)
- ack app (bypassed)

For Feedback type "nack" or negative acknowledgment:
- nack (locally consumed)
- nack pli (bypasssed)
- nack sli (bypassed)
- nack rpsi (bypassed)
- nack app (bypassed)

For Feedback type "cc" codec control:
- cc fir (bypassed)

The following, which are shown in the previous SDPs, are not supported:
- goog-remb
- [transport-cc](https://tools.ietf.org/html/draft-holmer-rmcat-transport-wide-cc-extensions) (this one looks promising BTW, but until we generate it we better remove the support from the SDP to avoid a ton of Feedback coming from Chrome.

**Comment by ibc** [comment]
Thanks! I will work on properly setting the supported values.

**Comment by ibc** [comment]
@jmillan I'm pretty sure you don't need it. However FYI "JSEP" draft also defines RTCP routing in [Appendix B](https://tools.ietf.org/html/draft-ietf-rtcweb-jsep-18#appendix-B);

**Comment by ibc** [comment]
Next actions:

1. Improve/implement proper generation timing for Receiver Reports.
1. Implement [draft-ietf-rmcat-gcc](https://tools.ietf.org/html/draft-ietf-rmcat-gcc) (Google Congestion Control Algorithm).
  * This also requires proper handling and mapping of RTP header extensions.

**Comment by ibc** [comment]
Next actions:

1. Improve/implement proper generation timing for Receiver Reports.
1. Implement [draft-ietf-rmcat-gcc](https://tools.ietf.org/html/draft-ietf-rmcat-gcc) (Google Congestion Control Algorithm).
   * This also requires proper handling of RTP header extensions.
1. Move the RTP buffer from the `RtpStream` class to the `RtpSender` (this is because, once simulcast/SVC is done, we will handle different sequence numbers for each receiver).
1. Regardless how many RTP packets we store in the `RtpStream` instance, don't satisfy NACK requests for packets older than 200ms (thanks @lminiero).
1. Locally generate REMB feedback messages to limit the sending bandwidth of remote senders (thanks @lminiero).
   * We may also process REMB from remote receivers and, in case most of them don't have a huge downlink bandwidth, we may ask the sender to limit its uplink traffic via REMB.

**Comment by ibc** [comment]
Note for the future: (S)AVPF spec mandates support for TMMBR which is a way to limit the sender bandwidth (REMB does also that), so we should support it.

**Comment by ibc** [comment]
Question: Let's imagine that mediasoup is gonna generate its own NACK and TMMBR and send them to remote senders. Which value should be placed into the `SSRC of packet sender` field of those feedback messages? Also, which `SSRC of packet sender` is currently set in Receiver Reports and Sender Reports generated by mediasoup?

Such a `SSRC of packet sender` is not signaled in the SDP (it may be signaled in ORTC or let the browser choose).

**Comment by jmillan** [comment]
On packets generated by RtpReceiver (ie: Receiver Report), we are setting a SSRC of 0. On packets generated by RtpSender (ie: Sender Report), we are setting the SSRC value that is being sent.

This is the way to go IMHO. Instead of hacking an invented SSRC on RTCP generated by receivers just set it to 0. The receiver of such RTCP has no way to check the SSRC anyway.

**Comment by ibc** [comment]
Nice and clear.

**Comment by ibc** [comment]
Habemus NACK generation in master branch: 553f4af308cba3f649d975bf12503910401dc631

**Comment by ibc** [comment]
Note regarding receipt of feedback messages (in [draft-ietf-rmcat-rtp-cc-feedback](https://tools.ietf.org/html/draft-ietf-rmcat-rtp-cc-feedback)):

>
   Compound RTCP packets contain a Sender Report (SR) packet and a
   Source Description (SDES) packet, and an RTP Congestion Control
   Feedback (RC2F) packet [I-D.dt-rmcat-feedback-message].  Non-compound
   RTCP packets contain only the RC2F packet.

So we must be ready to parse feedback messages also in compound packets. But maybe this is already done, is it @jmillan? :)

**Comment by jmillan** [comment]
> So we must be ready to parse feedback messages also in compound packets. But maybe this is already done, is it @jmillan? :)

Absolutely!

**Comment by ibc** [comment]
Cool, so are those two "packets" implemented?

* [RTCP XR Block for Reporting Congestion Control Feedback](https://tools.ietf.org/html/draft-dt-rmcat-feedback-message-01#section-3.1)
* [RTP/AVPF Transport Layer Feedback for Congestion Control](https://tools.ietf.org/html/draft-dt-rmcat-feedback-message-01#section-3.2)

**Comment by jmillan** [comment]
No, to yet.

**Comment by ibc** [comment]
I'm trying to figure out whether those "standard feedback messages" are already being sent by browsers or not... If so, it's up to the sender (which may be the *mediasoup* `RtpSender`) to implement any of the [existing sender side bitrate estimation algorithms](https://github.com/ibc/mediasoup/wiki/Congestion-Control#sender-side-bitrate-estimation-algorithms).

**Comment by jmillan** [comment]
>I'm trying to figure out whether those "standard feedback messages" are already being sent by browsers or not.

They are not being sent as far as I see.

**Comment by ibc** [comment]
So we basically can deal with [draft-alvestrand-rmcat-remb](https://tools.ietf.org/html/draft-alvestrand-rmcat-remb) (implemented by Chrome and Firefox) and [draft-holmer-rmcat-transport-wide-cc-extensions](https://tools.ietf.org/html/draft-holmer-rmcat-transport-wide-cc-extensions) (implemented by Chrome), both of them discontinued specs, right?

**Comment by jmillan** [comment]
yes 👍

**Comment by ibc** [comment]
And we are done here :)
